### ISSUE TITLE
WPAx: proposal for a new hash line format (successor of hccapx) 

### ISSUE THREAD

**Issue by ZerBea** [hashcat/hashcat#1816]
hccapx is a custom format, specifically developed for hashcat.
The structure of .hccapx is fix and we are not able to handle
new "attack verctors" like PMKID. Also we waste GPU time,
because we have to run hashcat on WPA-EAPOL-PBKDF2 (2500)
and WPA-PMKID-PBKDF2 (16800) even we have a PMKID hash
and an EAPOL hash using the same network name. If we combine
both hashmodes to a single hashline, we will get full advantage
of reause of PBKDF2 on both hash types. Therefore we need a
new and more flexible hash format.
Handle WPA-EAPOL-PBKDF2 (2500) and WPA-PMKID-PBKDF2 (16800)
and WPA-EAPOL-PMK (2501) and WPA-PMKID-PMK (16800) in a single hashline.

1
The following entries are mandatory to calculate the PMK
using PBKDF2:
ESSID

2
The following entries are mandatory to crack the hash
in hash mode WPA-PMKID-PBKDF2 (16800):
mac_ap
mac_sta
PMKID

3
The following entries are mandatory to crack the hash
in hash mode WPA-EAPOL-PBKDF2 (2500):
mac_ap
mac_sta
nonce_ap
EAPOL_sta

In case of hash modes WPA-EAPOL-PMK (2501) and
WPA-PMKID-PMK (16800), we do not need step 1.

In case of AKM-defined 4-weay handshakes (WPA3, SAE, FILS, EAP),
we do not need step 1, too.
https://hashcat.net/forum/thread-7717-post-42759.html#pid42759

Also we should use a hexified hashline (only hex values, especially for ESSID)
like we do it allready by 16800 hashlines. We get advantage to
handle this hash lines with simple bash commands instead
of several tools to retrieve informations about the binary hccapx.
Also hexified ASCII hash lines can be can be easily posted
in varius forums (hashkiller, antichat,...).

Putting this ideas into a single hashline, it could look like this one:

hashcat should run WPA-PMKID-PBKDF2 and WPA-EAPOL-PBKDF2 on this hashline
mac_ap * mac_sta * PMKID * ANONCE * ESSID * EAPOL

hashcat should run WPA-PMKID-PBKDF2 on this hashline
Received only a PMKID
mac_ap * mac_sta * PMKID * * ESSID*

hashcat should run WPA-EAPOL-PBKDF2 on this hashline
mac_ap * mac_sta * * ANONCE * ESSID * EAPOL

hashcat should run WPA-PMKID-PMK and WPA-EAPOL-PMK on this hashline
mac_ap * mac_sta * PMKID * ANONCE * * EAPOL

hashcat should run WPA-PMKID-PMK on this hashline
Received only a PMKID
mac_ap * mac_sta * PMKID * * * 

hashcat should run WPA-EAPOL-PMK on this hashline
mac_ap * mac_sta * * ANONCE * * EAPOL

Additional we can add a control field (*0000*) to 
allow hcxtools to parse informations to hashcat,
like:
message_pair
use only PMK mode
use only EAPOL mode
has is AKM-defined and requires SHA_384 mic calculation instead of SHA-256
...
example:
mac_ap * mac_sta * PMKID * ANONCE * ESSID * EAPOL * control_field

**Comment by magnumripper** [comment]
> only hex values, especially for ESSID

Also, IIRC when parsing the ESSID hex field we should store the length and **never** use `strlen()` on the resulting data. So an ESSID of `6d6167006e756d` should have length set to 7.

**Comment by ZerBea** [comment]
Hi magnum.
Are you sure, we need an additional ESSID len field in the new hash line?
For example:
 * 6d6167006e756d * 
The parser must do a validity check
1: count characters between the 2 stars: 14
2: ESSID len = chars /2
3: Do we have hex values (%2 ==0) and are they valid (0..9, a..f, A..F)

It's similar to the 16800 hash line ESSID (which doesn't contain an ESSID len field).

**Comment by ZerBea** [comment]
ESSID 
These SSIDs can be zero to 32 octets (32 bytes) long. 
The 802.11 standards prior to the 2012 edition did not define any particular encoding/representation for SSIDs, which were expected to be treated and handled as an arbitrary sequence of 0–32 octets that are not limited to printable characters. The IEEE 802.11-2012 defines a tag that the SSID is UTF-8 encoded and when interpreting could contain any non-ISO basic Latin characters within it. Wireless network stacks must still be prepared to handle arbitrary values in the SSID field.
That is a good reason to use hex values instead of trying to convert an ESSID to printable characters in hash lines.

EAPOL - we should allow oversized EAPOL len. Found this ones during wpa-sec analysis:
256, 258, 262, 270, 278, 288, 294, 306, 310, 322, 326, 330, 334,  342, 358, 370, 374, 386, 390, 406, 422, 438, 484, 500, 502, 510
$ hcxpcaptool -V --eapol-out=eapols caparchiv/2018*
$ cat eapols | awk 'BEGIN { FS = ":" } ; { print $NF }' | awk '{print length}' | sort | uniq
BTW:
I'm preparing to implement more analysis options to hcxtools. 

CONTROL FIELD:
Right now we use pcapng comment field to parse additional informations from attacker/dumper to the conversion tool.
It would be great if we can use a CONTROL FIELD to parse additional informations from the conversion tool to the hash cracker.
attack/dumptool --pcapng--> conversion tool --control field--> hash cracker

And it would be awesome if we can use the potfile to parse additional information from hash cracker to the data base (but this will we another proposal: improved potfile line): 
attack/dumptool --pcapng--> conversion tool --control field--> hash cracker  --potfile--> data base

Than, to complete the WPAx cycle,  we can think about a feed back from the data base to the attacker/ dumper:
data base --xxxxx--> attack/dumptool

**Comment by ZerBea** [comment]
We have no additional field for MIC and key ver, because we can retrieve them from EAPOL,
as well as the length.
Please keep  in mind that we must expect more than the the allready implemented length:
PMK_LEN_MAX 64 (32, as yet)
WPA_EAPOL_KEY_MIC_MAX_LEN 32 (16, as yet)

Also please notice, that we have a new key descriptor value (0=AKM-defined instead of 1=wpa1, 2=wpa2 , 3=wpa2 key ver 3) on AKM-defined handshakes.
Exampel is here:
https://hashcat.net/forum/thread-7717-post-42759.html#pid42759
Right now hcxpcaptool will set this value to 3 (AES-128-CNC), so that hashcat is able to run hash mode 2501 on that handshakes (currently that works on SAE, but it's a really, really ugly hack). In future times we can use the control field to parse AKM-defined data to the cracker.

**Comment by RealEnder** [comment]
Also we should not forget LE/BE/ap-less info in control field. How do you extract MIC and keyver from EAPOL, eg. gor example EAPOL value here: https://hashcat.net/wiki/doku.php?id=hccapx?

**Comment by ZerBea** [comment]
These are the 2 structs for an authentication frame and a wpa-key frame:

struct eapauthentication_frame
{
 uint8_t	version;
 uint8_t	type;
#define EAP_PACKET 0
#define EAPOL_START 1
#define EAPOL_LOGOFF 2
#define EAPOL_KEY 3
#define EAPOL_ASF 4
#define EAPOL_MKA 5
 uint16_t	len;
 uint8_t	data[1];
} __attribute__((__packed__));
typedef struct eapauthentication_frame eapauth_t;
#define	EAPAUTH_SIZE offsetof(eapauth_t, data)


struct wpakey_frame
{
 uint8_t	keydescriptor;
 uint16_t	keyinfo;
 uint16_t	keylen;
 uint64_t	replaycount;
 uint8_t	nonce[32];
 uint8_t	keyiv[16];
 uint64_t	keyrsc;
 uint8_t	keyid[8];
 uint8_t	keymic[16];
 uint16_t	wpadatalen;
 uint8_t	data[1];
} __attribute__((__packed__));
typedef struct wpakey_frame wpakey_t;
#define	WPAKEY_SIZE offsetof(wpakey_t, data)

So it's easy to lay the struct over the EAPOL frame, using bash commands.
Let's take the example from above and get EAPOL data:
$ hcxpcaptool -V sae4way.pcapng --eapol-out=example
reading from sae4way.pcapng
summary:                                        
file name....................: sae4way.pcapng
file type....................: pcapng 1.0
file hardware information....: x86_64
file os information..........: Linux 4.18.16-arch1-1-ARCH
file application information.: hcxdumptool 5.0.0
network type.................: DLT_IEEE802_11_RADIO (127)
endianess....................: little endian
read errors..................: flawless
packets inside...............: 15
skipped packets..............: 0
packets with GPS data........: 0
packets with FCS.............: 0
beacons (with ESSID inside)..: 1
probe requests...............: 1
probe responses..............: 1
association requests.........: 1
association responses........: 1
authentications (SAE)........: 4
EAPOL packets................: 7
EAPOL PMKIDs.................: 1
best handshakes..............: 1 (ap-less: 0)

Now let's get the MICs:
$ cat example | cut -c 189-220
c652368806f943802f8f57789e1a20d1
60e5524831cfaff7438ae1c365b9a9e4

an the key ver:
$ cat example | cut -c 40-40
8
8
8
a 
a

explanation:
8 = AKM-defined, pair wise key
a = WPA2, pair wise key (hcxpcaptool attack vector to downgrade to WPA2 - failed!)

Or we can add some options to the conversion tool to extract this informations, additional for a data base.

Please keep in mind that we use only single EAPOL data for this tests, not a complete, new hash line.
If we need fix positions, we should change position from * ESSID * EAPOL * to * EAPOL * ESSID *

**Comment by ZerBea** [comment]
Maybe it's necessary to increase control field size ( * 00000000 * ) to parse more informations to cracker. Current informations:
nonce error corrections needed on BE
nonce error corrections needed on LE
nonce error corrections needed on BE/LE
no nonce error corrections needed (hash is a result of an AP-LESS attack vector - we should split this)
Information of message pair is not necessary, because we can retrieve this from EAPOL data

new :
use AKM-defined MIC / PMKID calculation (SHA_192, SHA384, SHA_512, AES-128-CNC for SAE/FILS/WPA3)
use PMK len 64
PMKID received from AP or PMKID received from CLIENT (we got the first one of this kind)
no nonce error corrections needed
hash is a result of an AP-LESS attack vector

**Comment by magnumripper** [comment]
> Hi magnum.
> Are you sure, we need an additional ESSID len field in the new hash line?

No, I meant when *parsing* that field we need to look at its length instead of doing a strlen on the result. So hashcat and JtR should have a len field in whatever struct is holding the data, but the input field doesn't need it.

**Comment by ZerBea** [comment]
I am a little blinded, because I analyzed plenty of cap files during the last week to test the new hash line.  Reuse of PBKDF2 over EAPOL and PMKID makes cracking process really, really fast.
The same applies to reuse of PMK over EAPOL and PMKID.
So thanks for your explanation.
BTW:
From my point of view (hcxtools) a concerted hash line for hashcat and JtR would be fantastic!!!

**Comment by ZerBea** [comment]
Some early words about the potfile.
It would be great to retrieve a potfile line, containing main informations about the result, like this one: 
mac_ap * mac_sta * PMK * PMKID * control field * ESSID (in hex) : PSK
Parsing this kind of information, the hash cracker (hashcat) become part of the whole analysis process. It can add additional informations (for example: how the PSK was retrieved) to the control field and parse it to the data base.

**Comment by ZerBea** [comment]
More words about the potfile:
Right now, hashcat use a md5_64 hash over EAPOL (and some other fields, like mac, anonce, ESSID) to make a hash unique and to remove duplicates. Unfortuneatelay an EAPOL message contains several values of counters and many flags, which changes from hash to hash. So hashcat calculates on every hash a new md5_64. A better way is to use the calculated PMK to make the network (and not the hash) unique. That will work, even if the incomming new hash has no ESSID or a damaged ESSID. That is a real benefit of a combined hash line (2500, 2501, 16800, 16801) and a combined potfile line ((2500, 2501, 16800, 16801). Also it will keep the potfile small.

**Comment by ZerBea** [comment]
How can we help?

**Comment by jsteube** [comment]
New format in hashcat with hash-mode 22000 has been added. hcxpcapngtool also have support added.

The new format is:

```
SIGNATURE*TYPE*PMKID/MIC*MACAP*MACSTA*ESSID*ANONCE*EAPOL*MESSAGEPAIR
```

Some notes:

- SIGNATURE = "WPA"
- TYPE = 01 for PMKID, 02 for EAPOL, others to follow
- PMKID/MIC = PMKID if TYPE==01, MIC if TYPE==02
- MACAP = MAC of AP
- MACSTA = MAC of station
- ESSID = ESSID
- ANONCE = ANONCE
- EAPOL = EAPOL (SNONCE is in here)
- MESSAGEPAIR = Bitmask:

```
0: MP info (https://hashcat.net/wiki/doku.php?id=hccapx)
1: MP info (https://hashcat.net/wiki/doku.php?id=hccapx)
2: MP info (https://hashcat.net/wiki/doku.php?id=hccapx)
3: x (unused)
4: ap-less attack (set to 1) - no nonce-error-corrections necessary
5: LE router detected (set to 1) - nonce-error-corrections only for LE necessary
6: BE router detected (set to 1) - nonce-error-corrections only for BE necessary
7: not replaycount checked (set to 1) - replaycount not checked, nonce-error-corrections definitely necessary
```

All fields are encoded in hex except SIGNATURE.

See https://github.com/hashcat/hashcat/blob/master/src/modules/module_22000.c#L743-L804 for details.

Also note hashcat mode 22000 is backward compatible and supports loading for hccapx as well as hashlines formated for mode 16800.

**Comment by magnumripper** [comment]
So you opted to use `:` as field delimiter instead of the proposed `*`... Unfortunately JtR won't support this ever, for good reasons 😢

**Comment by ZerBea** [comment]
That shouldn't be a big problem, because hcxpcapngtool will support JtR, too. PMKID part is ready for JtR and now I'm working on the EAPOL part, which should be added today. 
https://github.com/ZerBea/hcxtools/blob/master/hcxpcapngtool.c#L331
https://github.com/ZerBea/hcxtools/blob/master/hcxpcapngtool.c#L290
Now both both hash crackers doing reuse BPKDF2 over PMKID and EAPOL, so we can use a single file on PMKID and EAPOL (-o hashcat -j john). That is a big advantage.
Unfortunatley JtR EAPOL format has three big disadvantages:
- internally is relies on old (deprecated and fixed) hccap struct.
- the EAPOL frame is "compressed" and it must be "decompressed" before we can retrieve information from the EAPOL frame. So bash tools will not work out of the box on them.
- some field content is repeated: MAC_AP, ESSID, WPA (key ver) which is already in the EAPOL frame, authorized information - which can be done in a single field.

BTW:
An analysis of wpa-sec recovered PSKs and ESSIDs showing that more PSKs and ESSIDs contain * than : .
To make it more bash tool use able, PSKs and ESSIDs containing the delimiter must be hexified.  That means more hexification on * .
But the delimiter is only cosmetics and can be changed by bash tool (eg awk, sed) in a fast way.

**Comment by jsteube** [comment]
It's not too late to switch from `:` to `*` as separator character as code was not tagged as release so far. I'm OK changing it. Not sure if it's a problem in hcxpcapngtool but I don't think so?

**Comment by ZerBea** [comment]
No, it isn't a problem for hcxtools/hcxpcaptool. And it shouldn't be a problem fro the hash cracker, too.
Think about the PMKID line.
If the user decided to run  on a different hash cracker he can do simple sed cmd to change  for his desired hash cracker.
A real big problem is the compressed EAPOL frame in JtR. If I would like to do nonce correction (not build in in JtR), I have to repack the complete hash line for JtR instead of grabbing the ANONCE field, changes a single byte and writing it back.
For me, hashcat and JtR are the main hash cracking tools and hcxpcapngtool will support both of them.

**Comment by RealEnder** [comment]
I'm left with the impression @magnumripper has no objections against the hashline itself, which is generally easier to work with, only the separator. If this is the showstopper, lets change it. For JtR the only difference left will be, that we still have to do nonce correction with multiple hashlines, but this is not a problem at all IMO.

**Comment by ZerBea** [comment]
Maybe using both delimiters : on hashcat and * on JtR isn't a so bad idea, because, if we meet a hashline with delimiter *, we can expect NC overhead inside, while the : is clean.

**Comment by magnumripper** [comment]
> It's not too late to switch from `:` to `*` as separator character as code was not tagged as release so far. I'm OK changing it. Not sure if it's a problem in hcxpcapngtool but I don't think so?

That would be awesome - I'd implement the very same format in JtR as soon as I get time.

> Maybe using both delimiters : on hashcat and * on JtR isn't a so bad idea, because, if we meet a hashline with delimiter *, we can expect NC overhead inside, while the : is clean.

True but it would be much better to drop that stupid hack of mine and have JtR behave more like hashcat - with some kind of nonce-fuzzing option.

**Comment by ZerBea** [comment]
For sure, running the same hash line on hashcat and JtR makes life much easier, for hcxtools, too.
The new hash line is really good. No overhead at all, no double fields, extreme flexible. It shouldn't fail because of a stupid delimter. 6.0.0 isn't released, so we can change it. As Jens said, it's not too late.

**Comment by ZerBea** [comment]
And as long as the new hash line (hashcat and JtR) haven't arrived in the major distros, I'll support
- the new hash line, of course (it is really good)
- hccapx
- hccap ( unfortunately, there are still many users, running this ancient format, especially on old hardware)
- JtR old format (at the moment still JtR current format - well, it's really a stupid hack, like hccap, too)

**Comment by magnumripper** [comment]
> * MESSAGEPAIR = Bitmask:
> 
> ```
> 4: ap-less attack (set to 1) - no nonce-error-corrections necessary
> 5: LE router detected (set to 1) - nonce-error-corrections only for LE necessary
> 6: BE router detected (set to 1) - nonce-error-corrections only for BE necessary
> 7: not replaycount checked (set to 1) - replaycount not checked, nonce-error-corrections definitely necessary
> ```

I suppose there might be situations where bit 5 & 6 are both unset and we need to fuzz for *both* LE and BE but I'd probably reject lines where bit 4 & 7 are both set. BTW is there ever any reason to do nonce fuzzing when bit 7 is unset? If not, that bit (**or** bit 4) is superflous.

**Comment by magnumripper** [comment]
A corresponding pot line in JtR will only contain hexified ESSID and calculated PMK, something like
```
WPA*00*PMK****ESSID***:password
```
That is: TYPE is set to 00, next field is calculated PMK (32 bytes / 64 hex digits) and all other fields but ESSID are empty (or 00?).

@jsteube @ZerBea this means we'd reserve TYPE 00 for this use, OK?

**Comment by ZerBea** [comment]
That is correct:
- bit 5 & 6 are both unset and we need to fuzz for both LE and BE
bit 4 & 7 are both set
- that should not happen, because we check RC and ANONCE. If they match to hcxdump attack value (retrieved from pcapng comment field), we set bit 4
- we set bit 7 if user choose option nonce-error-corrections=<digit> and enter a value here.
- but in every case, we override this user  selection on a hcxdumptool ap-less attack vector: https://github.com/ZerBea/hcxtools/blob/master/hcxpcapngtool.c#L343
reason to do nonce fuzzing when bit 7 is unset
- no. The automatic handshake detection should work fine. If a desired handshake isn't converted, the user can do a re-run of hcxpcapngtool with activated nonce-error-corrections or convert all handshakes (do-not-clean option). We unset this bit 7 if we assume we have a good handshake from which the PSK can be recovered) and that will save us GPU time.

If hashcat is ready (pushed) for * delimiter, I'll push hcxpcapngtool, too.
If JtR is ready for the new hash line, I move old format output from -j to --john.

**Comment by ZerBea** [comment]
Normally we don't need this
WPA * 00 * PMK **** ESSID *** password

Potfile entry:
PMK * ESSID_IN_HEX * PSK hexified if delimiter * is inside
example:
d915ebf5775c1e88a9855439511df2a5a2d24eb77803a9da1ef19acbab439b3f * 64656661756c74 * 00000000
That keep the potfile smal and machine readable. Also this are exactly the (static) parameters of PBKDF2: PMK * SALT * PSK

additional you can add a user friendly output (not for potfile)
MAC_AP * MAC_CLIENT * ESSID * PSK
112233445566 * aabbccddeeff * default * 00000000

BTW:
Unfortunately I'm running into an issue, when I try to comment an example of the hash line, because
this comment box interpreted both * and : as control characters.

**Comment by magnumripper** [comment]
> A corresponding pot line in JtR will only contain hexified ESSID and calculated PMK, something like
> 
> ```
> WPA*00*PMK****ESSID***:password
> ```
> 
> That is: TYPE is set to 00, next field is calculated PMK (32 bytes / 64 hex digits) and all other fields but ESSID are empty (or 00?).

No, wait. That will be impossible to match when just running `--show`. I probably need to keep the input and tuck the calculated PMK in an *additional* field in the end.

An input line of
```
WPA*<TYPE>*<PMKID/MIC>*<MACAP>*<MACSTA>*<ESSID>*<ANONCE>*<EAPOL>*<MESSAGEPAIR>
```
Will have a pot line of 
```
WPA*<TYPE>*<PMKID/MIC>*<MACAP>*<MACSTA>*<ESSID>*<ANONCE>*<EAPOL>*<MESSAGEPAIR>*<CALCULATED_PMK>:password
```

> Potfile entry:
> `PMK * ESSID_IN_HEX * PSK hexified if delimiter * is inside`

No, that last `*` isn't part of the "hash" so it will be using the field limiter of `:` (and wont need hexifying even if it contains `:` because in a JtR potfile, only the *first* `:` is the delimiter - but we'd need to hexify if PSK contains a newline, null or other strange things).

> example:
> `d915ebf5775c1e88a9855439511df2a5a2d24eb77803a9da1ef19acbab439b3f*64656661756c74*00000000`

This will be
`d915ebf5775c1e88a9855439511df2a5a2d24eb77803a9da1ef19acbab439b3f*64656661756c74:00000000`

> That keep the potfile smal and machine readable. Also this are exactly the (static) parameters of `PBKDF2: PMK * SALT * PSK`

Still, we can't match such pot entry to an input hash when running `--show`. How would we?

**Comment by ZerBea** [comment]
That was a misunderstanding. I described the hashcat potfile and hashcat user output.
There is absolutely no need for JtR to do it in the same way.
It is only important, that the potfile can be edited by simple bash tools.

**Comment by magnumripper** [comment]
I still can't understand how hashcat would match an input line to such a pot line? For running `--show`, I mean.

**Comment by ZerBea** [comment]
I think Atom have spent many hours to code this part. And I spent many hours to explain it to him (paying the price for PBKDF2 once and writing the result to potfile - than using the result as "kind of a rainbowtable" on every matching PMK-SALT-PSK combination). That is very fast, and produces no overhead.

**Comment by ZerBea** [comment]
But now, he can explain it much better than I can do. Now, his solution is brilliant and much, much better than my first ideas.

**Comment by ZerBea** [comment]
Just talked with Atom. We'll switch to * as delimiter. I'll push hcxtools update, soon.

**Comment by magnumripper** [comment]
OK I think I see how to match pot-entries at load time. I wouldn't call it "no overhead" but it's still a brilliant idea for avoiding lots of PBKDF2 work later. I can't wait to try this out!

**Comment by ZerBea** [comment]
For sure, doing this in that way produce some overhead.
But this, done all all new incoming hashes is working fine (hashcat example:
cut -c -64 hashcat.archiv.potfile | hashcat -m 16801 --logfile-disable -w 3 --remove --potfile-disable -o "/tmp/hashcat.pmk" new.16800"
cut -c -64 hashcat.archiv.potfile | hashcat -m 2501 --logfile-disable -w 3 --remove --nonce-error-corrections=32 --potfile-disable -o "/tmp/hashcat.pmk" new.hccapx"
will remove all recovered PSKs in a very short time.

**Comment by ZerBea** [comment]
Another important information about hashmode 22001 / PMK:
We know 4 key versions:
0 AKM defined
1 WPA1
2 WPA2
3 WPA2 key version 3
Some AKM defined / EAP defined authentications run 4way handshake, too and we are able to verify the used dynamic PMK (if we have one). I wrote about this here:
https://hashcat.net/forum/thread-7717-post-42759.html#pid42759
and made a pull request here - this pull request must be removed in 22001
https://github.com/hashcat/hashcat/pull/1986
Running the old format hcxpcaptool detected that and moved key version field in hccapx struct from 0 to 2. That will not work any longer on hashmode 22001.
Now, if the user requested hashmode 22001/PMK and key version field in EAPOL frame is 0, hashcat and JtR must handle this, as WPA2 HMAC(EVP_sha1()), otherwise we are not able to verify this PMK. hcxpcapngtool will convert this handshakes only on request: --ignore-ie, because PBKDF2 is useless, here.

**Comment by ZerBea** [comment]
By this commit I marked hccapx, hccap and current JtR format as deprecated/old:
https://github.com/ZerBea/hcxtools/commit/7b99f59c2e2f4507cbdc77a7e67c4b3248b32241
Next push will be the move from delimiter : to *

**Comment by ZerBea** [comment]
Pushed update. Now running * as delimiter on new hashline (PMKID and EAPOL). Conversion from pcapng/pcap/cap (gz compressed, too) should work as expected and you can use this for testing.
GPS (NMEA) output, as well as output of wordlists (retrieved from traffic) and TACACS+, MD5 challenge, LEAP, PEAP, IMEI detection is still missing in that version. I'm going to re-add this features step by step, before removing deprecated hcxpcaptool from git.
Priority and goal is that it work with hashcat and JtR.

**Comment by RealEnder** [comment]
Guys, while here, what about extending the INFO field? It's nice to have a placeholder for GPS coordinates and additional seed words for JtR. Every tool can put there its own semantics. This can just be additional/optional COMMENT field after the last one.

**Comment by ZerBea** [comment]
That should be no problem, because, now, we have an HEX-ASCII line. We should be able to add this fields at the end of the line. hashcat and JtR can either ignore them or the user has to remove this field(s) by hand, before feeding the hashline to hashcat and/or JtR.
I can add everything here. Just tell me the fields.

**Comment by ZerBea** [comment]
Started support for new hash line format on whole hcxtools suite:
https://github.com/ZerBea/hcxtools/commit/853a353f061cc55cb7c8a59ba6d426db660bf3f3
that we can convert old files to new hash line, by running/cascading hcxtools

**Comment by magnumripper** [comment]
> Guys, while here, what about extending the INFO field? It's nice to have a placeholder for GPS coordinates and additional seed words for JtR. Every tool can put there its own semantics. This can just be additional/optional COMMENT field after the last one.

Not sure how hashcat would cope with it but JtR always support the "good" old /etc/passwd syntax (that's why we can't have any `:` **within** the hash). So given a "hash" of
```
b1d3b1f2109b841774c2fbec368d646e*08606ebb2cb8*b8ee65ecf96a*41535553
```
(old-style, but that doesn't matter for this discussion)

We can add "login" (== ESSID as text if possible), "uid", "gid", "gecos" and optionally "home directory" fields á la `/etc/passwd` like this
```
ASUS:b1d3b1f2109b841774c2fbec368d646e*08606ebb2cb8*b8ee65ecf96a*41535553:::Any info in this field (the GECOS field) will be used as seeds
```
Note that `:` is the separator for *input line fields*, separating eg. the "login" from the "hash". On the other hand `*` is the separator for *fields within the hash*.

But I digress. What I mean is we can put, for example, street names derived from GPS info in the GECOS field. It can then be used as seed specific for that hash only.

**Comment by jsteube** [comment]
In hashcat you can disable this /etc/passwd functionality per module:

https://github.com/hashcat/hashcat/blob/master/src/modules/module_22000.c#L287-L292

**Comment by magnumripper** [comment]
@jsteube I noticed the pot entries are still using `:` between key and ESSID. Not a big deal, but if you could change that one too to `*` we'd be pot-file compatible as well.

**Comment by jsteube** [comment]
@magnumripper ok done https://github.com/hashcat/hashcat/commit/daaf5d365c0cc1ed1564cf41683cd66d45a4c07a
@ZerBea fyi

**Comment by magnumripper** [comment]
Cool, thanks!

**Comment by ZerBea** [comment]
We should change this potfile entries for 250x/1680x, too, because we're running all on the same potfile.

**Comment by jsteube** [comment]
ok done https://github.com/hashcat/hashcat/commit/20ef9725ef30cb992d5ba422ea6e2cf2e3e3683c

**Comment by ZerBea** [comment]
That will make life easier for the transition time from 250x/1680x to 2200x
Great, thanks.

**Comment by ChemicalNRG** [comment]
maybe a stupid question, but is it possible to add a potfile for tried but failed crack's?
And maybe with new hashline in extra info field.
So if i did try to crack a pmkid with 8 or 9 digits, hashcat will know i already tried that one. Maybe in a separate file per charset or wordlist in hex format to keep it fast. If i start an attack with ?d?d?d?d?d?d?d?d?d it will check 3f643f643f643f643f643f643f643f643f64.failed or something if AP is allready in there so not only found keys but also tried ones will be skipped.

**Comment by philsmd** [comment]
@ChemicalNRG What you describe in here is basically exactly what the "hashcat brain" does, see https://hashcat.net/forum/thread-7903.html (scroll to "Major Feature: The hashcat brain").

This of course only works for slow hash types, otherwise there is way too much data to keep (in hashcat's brain). I think this discussion is very of topic here and you should try to find some answers over at the hashcat forum (e.g. start with the link above and find more forum posts from there).
